functions/internal/util/util.go

```diff
@@ -176,3 +181,53 @@ func ReadCryptoRandBytes(b []byte) {
 		panic(fmt.Errorf("could not read random bytes: %v", err))
 	}
 }
+
+// CheckHTTPS retrieves the scheme from the X-Forwarded-Proto or RFC7239
+/*
+We do this because in the function running in a cloud container the SSL termination
+has likely happened upstream so we need to check the headers to reject HTTP only.
+Requests on GCE contain both of these headers and anything supplied by the client is
+overwritten. Locally in development mode we don't use HTTPS so the client should send
+one of these headers.
+*/
+
+var (
+	// De-facto standard header keys.
+	xForwardedProto = http.CanonicalHeaderKey("X-Forwarded-Proto")
+	forwarded       = http.CanonicalHeaderKey("Forwarded") // RFC7239
+	
+	protoRegex 		= regexp.MustCompile(`(?i)(?:proto=)(https|http)`)
+)
+
+func CheckHTTPS(r *http.Request) StatusError {
+	var scheme string
+
+	// Retrieve the scheme from X-Forwarded-Proto.
+	if proto := r.Header.Get(xForwardedProto); proto != "" {
+		scheme = strings.ToLower(proto)
+	} else if proto = r.Header.Get(forwarded); proto != "" {
+		// match should contain at least two elements if the protocol was
+		// specified in the Forwarded header. The first element will always be
+		// the 'proto=' capture, which we ignore. In the case of multiple proto
+		// parameters (invalid) we only extract the first.
+		if match := protoRegex.FindStringSubmatch(proto); len(match) > 1 {
+			scheme = strings.ToLower(match[1])
+		}
+	}
+
+	if scheme != "https" {
+		var err = statusError {
+			code: http.StatusUpgradeRequired,
```

Comment:
Can't believe I get to say this, but let's use `http.StatusTeapot`. I would also add the following comment explaining why we chose this code:

```go
// We want to ensure that clients always use HTTPS. Even if we don't
// serve our API over HTTP, if clients use HTTP, they are vulnerable
// to man-in-the-middle attacks in which the attacker communicates
// with our service over HTTPS. In order to prevent this, it is not
// sufficient to simply auto-upgrade to HTTPS (e.g., via a redirect
// status code in the 300s). If we do this, then code which
// erroneously uses HTTP will continue to work, and so it might get
// deployed. Instead, we have to ensure that such code breaks
// completely, alerting the code's developers to the issue, and
// ensuring that they will change the code to use HTTPS directly.
// Thus, we want an error code with the following properties:
//  - Guaranteed that smart clients (such as web browsers) will not
//    attempt to automatically upgrade to HTTPS
//  - Doesn't have another meaning which might cause developers to
//    overlook the error or lead them down the wrong path (e.g., if
//    we chose 400 - bad request - they might go down the path of
//    debugging their request format)
//
// For these reasons, we choose error code 418 - server is a teapot.
// It is as unlikely as any other error code to cause the client to
// automatically upgrade to HTTPS, and it is guaranteed to get a
// developer's attention, hopefully getting them to look at the
// response body, which will contain the relevant information.
```

(this has already been wrapped to 80 columns when intended by three tabs)

Reply:
😂Achievement unlocked! First use of HTTP Status Code 418 "I'm a teapot". 🍵